### PR TITLE
Introduce Fastlane to Reaxit

### PR DESCRIPTION
**Pull request by se-bastiaan**
### Summary

Add fastlane configuration to make it easier to release the application to the public.

### How to test
Steps to test the changes you made:
1. Checkout all possible Fastlane actions, they are in the README
2. Check the CI builds for Android & iOS

### Fastlane

This PR contains a [Fastlane configuration](https://fastlane.tools) that you can use to build and deploy the app for Android and iOS.
To use Fastlane follow these steps:
 1. To be able to start you need [an installation of Ruby](https://www.ruby-lang.org/en/documentation/installation/)
 2. The first time run `bundle install`
 3. Then use fastlane by running `fastlane <platform> <command>`.

In the case of iOS we use [Fastlane Match](https://docs.fastlane.tools/actions/match/) to setup the certificates required to build apps.
It can help to setup a `.env` file in the `fastlane` folder for your convenience.

| Command           | Description                     | Platforms    |
| :---------------- | :------------------------------ | :----------- |
| deploy_adhoc      | Create a release or AdHoc build | Android, iOS |
| deploy_internal   | Create a release and deploy to the internal Play Store track or Testflight | Android, iOS |
| deploy_beta       | Create a release and deploy to Play Store beta or Testflight with external testers | Android, iOS |
| deploy_production | Create a release and deploy to Play Store or App Store | Android, iOS |
| match             | Get the certificates to sign iOS apps | iOS |

### CODE REVIEW

**Comment by DeD1rk** [comment]
Fastlane will need to pass secrets and other configuration (host, oauth identifier and secret, sentry dsn, and possibly more later) through dart-defines. That is, `flutter build` should get arguments `--dart-define=VAR_NAME="VALUE"` for each required variable in [lib/config.dart](https://github.com/svthalia/Reaxit/blob/57024e92a5a1d993dc95878d906c5b5b1c09834f/lib/config.dart)

**Comment by se-bastiaan** [comment]
Deploy still needs work, and I need to add environment variables for the url and keys.

File: fastlane/Fastfile-iOS [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        password: ENV["KEYCHAIN_PASSWORD"],
+        default_keychain: true,
+        unlock: true,
+        timeout: 3600,
+        lock_when_sleeps: false
+      )
+      app_store_connect_api_key(
+        key_id: "H37PU3Q6JX",
+        issuer_id: "69a6de95-e27f-47e3-e053-5b8c7c11a4d1",
+        key_filepath: "../app-store.p8",
```
Comment (se-bastiaan) [inline fastlane/Fastfile-iOS:15]:
```suggestion
        key_filepath: "./app-store.p8",
```

File: fastlane/Fastfile-iOS [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        lock_when_sleeps: false
+      )
+      app_store_connect_api_key(
+        key_id: "H37PU3Q6JX",
+        issuer_id: "69a6de95-e27f-47e3-e053-5b8c7c11a4d1",
+        key_filepath: "./app-store.p8",
+      )
+    end
+    sh("flutter pub get")
+    cocoapods(use_bundle_exec: false)
```
Comment (se-bastiaan) [inline fastlane/Fastfile-iOS:19]:
```suggestion
    cocoapods(
      use_bundle_exec: false,
      podfile: './ios'
    )
```

**Comment by DeD1rk** [comment]
Awesome, will review tomorrow probs

**Review by DeD1rk — COMMENTED** [review]
For me `yarn fastlane:install` does not work:
```
yarn run v1.22.10
error Couldn't find a package.json file in "/Users/dirkdoesburg/Technicie/Reaxit"
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

However, I can use `fastlane <lane>`. I'm not sure whether that is normal, but it should probably be documented in the README. Additionally, I think it would be good to refer/guide towards the .env file that we need to create from the README (especially, where do you find the App Store Connect Team ID and Developer Portal Team ID?). I can't get fastlane to work for iOS locally without these values. 

EDIT:
I got it working now. Are the 'App Store Connect Team ID' and 'Developer Portal Team ID' not needed? It seems like I can leave them out of the .env file. Anyway, we should document exactly how to set it up, i.e. decode and place the app store api key in `./app-store.p8`, create the .env (what files in the password repo contains those values), run `fast lane match`.

**Comment by se-bastiaan** [comment]
@DeD1rk 

Ah sorry, I forgot to update the PR description. The right instruction was in the README.
Technically one does not need to use fastlane for local development. I mostly made it for deployment and that can always happen in CI.

But you're right that I should document those steps. Based on [the documentation that I just read](https://docs.fastlane.tools/advanced/Appfile/) some values may indeed not be necessary. I probably took them from some tutorial or they were added during the initial init process.

We could also choose to just hardcode the values to the Thalia information. The email address for the login and the team id are not secret in any way I believe. And while typing this I realise I have already hardcoded the app store key information.

**Comment by DeD1rk** [comment]
I see, the README looks better indeed. I think hardcoding the not-so-secret stuff is probably a good idea. We can then refer simply to the match password and the app store api keys. I did have to make the `app-store.p8` file myself though, and it isn't in the README, so that should be documented. Finally, the sentry DSN is not passed yet (I guess I should probably also add a configuration for the build number or so? So I could do that together later).

**Comment by se-bastiaan** [comment]
You didn't have to make the .p8 file yourself, just have to document that it is in the thaliapp-passwords repo.

I did set the environment variables for Sentry and assumed that the SDK worked the same and would grab those values, but I now see in the [documentation](https://docs.sentry.io/platforms/flutter/#configure) that they already expect values using --dart-define too.

I am not sure what you mean by the build number?

**Comment by DeD1rk** [comment]
> just have to document that it is in the thaliapp-passwords repo. 
exactly

> they already expect values using --dart-define too
Thats very interesting. I recall reading that sentry tries to read those environment variables at runtime if they're not passed in code, but didn't know they built in using dart-defines already, so I already added them to `config.dart` as well. 

> I am not sure what you mean by the build number?
I think it would be nice to let sentry know the exact version number (e.g. 3.0.0-39), but I see they already take `SENTRY_RELEASE` through dart-defines, so that works perfectly.

Then if you add the bit of documentation and dart-defines this is ready to merge 🤩.

**Comment by se-bastiaan** [comment]
Apart from that I do not have the release versions in fastlane the [documentation ](https://docs.sentry.io/platforms/flutter/configuration/releases/) mentions
> If no release name is set for Android and iOS, the SDK creates a default combined of packageName, versionName and versionCode, for example my.project.name@2.3.12+1234.

So I don't think I have to add the release somewhere.

**Comment by DeD1rk** [comment]
Nice documentation and works perfectly for iOS like this. Only for android I get 
```
> A failure occurred while executing com.android.build.gradle.internal.tasks.Workers$ActionFacade

   > com.android.ide.common.signing.KeytoolException: Failed to read key appsigning from store "/Users/dirkdoesburg/Technicie/ThaliApp-passwords/thaliapp.keystore": Keystore was tampered with, or password was incorrect
```

With thaliapp.keystore fresh from the passwords repo and the storePassword and keyPassword set as instructed.

**Review by DeD1rk — APPROVED** [review]

**Pull request merged**